### PR TITLE
Add support for PHP 8

### LINKED ISSUE #15: PHP 8.0 compatibility

### PR DESCRIPTION
**Pull request by dshoreman**
## Description and Background

With the recent changes in ChrisB9/php8-xdebug that fix xdebug options, update composer and (afaict) remove the profiler that caused errors previously, there seem to be only two errors remaining in the tests on PHP 8.

Both errors are with the `assert()` lines, which throw an `AssertionError` in v8, but only trigger a warning (converted to `ErrorException`) when ran under v7. This is caused by a change in php8 which now enables the `assert.exception` option by default.

I considered a couple methods, but simply enabling the option by default seems to be the cleanest route. The option itself was added in v7, so this change just makes things consistent across both versions.

## How Has This Been Tested
Existing tests fixed, ran with `docker-compose run php composer test && docker-compose run php8 composer test`.

Psalm is happy, but the infection tests are currently still failing as a result of xdebug connection issues.

#### Checklist

- [x] My code follows the PSR coding guideline.
- [x] I have updated the `documentation` accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added `tests` to cover my changes.
- [x] All new and existing tests passed.

Resolves #15

### CODE REVIEW

**Comment by dshoreman** [comment]
The grumphp-config Composer dep was blocking Travis build due to lack of composer-2 support, so I've updated it to 4.0.1 which adds compatibility for it. Now that it's running again, it looks like the errors I'm getting locally with the mutation tests (xdebug could not connect to debugging client) are also happening in Travis, albeit only for php nightly.

I've tried updating the `EXPOSE` line from `9000` to `9003` in the php8-xdebug dockerfile and rebuilding from that local image, but that made no visible difference. Running `phpinfo` shows the settings from php8-xdebug (eg `mode=profile,debug,develop` etc) *are* in fact being applied so *that* seems fine at least... though I'm not really sure where to go from there.

**Comment by ChrisB9** [comment]
Thanks alot for your effort.
I think a major issue is, that the dependencies are not all PHP 8.0 ready and that may cause troubles on all fronts.
I might actually already merge this because the tests are running and the only reason why infection doesn't work is due to external factors. Maybe though we could first extend the dependencies here or extract that into a second pull request? Such as phpunit ^9.0, psalm ^4.0, infection ^0.19 and php-invoker ^3.0. (I know it will fail thanks to grumphp and webmozart/glob but both will add PHP 8.0 support soon.

**Comment by dshoreman** [comment]
I've just updated phpunit to 9.x and (so far) it's still green, although there are some new warnings now:

```
[snipped xdebug errors - see below]
PHPUnit 9.4.2 by Sebastian Bergmann and contributors.

Warning:       xdebug.mode=coverage has to be set in php.ini
Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

Error:         PHP extension pcntl is required for enforcing time limits
```

Running the migration command has kindly removed most line breaks so I gotta fix that before I can diff them, but with any luck we won't need to adjust *too* much.

Assuming the phpunit stuff goes well, I'll push that up then see if there's time to look at the rest before I pass out

---

For reference, these are the xdebug warnings from php itself:

```
Creating shellcommandbuilder_php8_run ... done
Xdebug: [Step Debug] Could not connect to debugging client. Tried: internal:9003 (fallback through xdebug.client_host/xdebug.client_port) :-(
Xdebug: [Profiler] File '/tmp/debug/cachegrind.out.1' could not be opened.
> vendor/bin/phpunit -c tests/phpunit.xml --testdox --color=always
Xdebug: [Step Debug] Could not connect to debugging client. Tried: internal:9003 (fallback through xdebug.client_host/xdebug.client_port) :-(
Xdebug: [Profiler] File '/tmp/debug/cachegrind.out.12' could not be opened.
```

**Comment by ChrisB9** [comment]
Regarding the XDebug error. It seems like this error happens because XDebug cannot establish a connection to PHPStorm.
https://3.xdebug.org/docs/step_debug#log
I changed the XDebug default settings disabling debug by default.
https://github.com/ChrisB9/php8-xdebug/pull/8/files

**Comment by dshoreman** [comment]
That would make sense as I don't have PHPStorm installed :D  
Thanks for the update. I'll pull the new docker bits later and give it a try, but that might just be all the warnings taken care of.

The "cachegrind.out" and /tmp/debug errors can be solved easily a couple ways, depending which you prefer:

1. Add `./tests/xdebug:/tmp/debug` (or similar) to `volumes:` list in ShellCommandBuilder/docker-compose.yml
   * How I fixed it locally (for now), seems to do the job
   * The debug dir requires 775 permissions, so it'd need to be tracked in git
   * To avoid unecessary cruft, `./tests/xdebug/*` would be added to .gitignore

2. Same as above, but in the php8-xdebug repo

3. `RUN mkdir /tmp/debug && chmod 775 $_` somewhere in php8-xdebug/php-dev-8.dockerfile

Option 3 *could* be the better route? But I'm not sure if other tools need to access the profiling data. If they do then we might need to go with option 1 regardless, in which case I can add it to this PR.

The only other question is that of the PHP minimum version. Phpunit requires `>=7.3` and infection requires `^7.4 || ^8`, but for now since both are dev-deps I think we'd be able to stay on 7.2 to avoid also updating that container. With 7.2 at EOL though, bumping that to 7.3 or .4 could be something for the future.

**Comment by ChrisB9** [comment]
I actually added this line too 
`RUN mkdir /tmp/debug && chmod 775 $_ somewhere in php8-xdebug/php-dev-8.dockerfile`

**Comment by dshoreman** [comment]
It turns out the `chmod` isn't needed when it's created directly in php-dev-8.dockerfile. I should've tested that beforehand, sorry! Good news though is everything's so far running smooth from what I can see.

I've updated phpunit and infection locally and both are now running without issue.  
Infection flagged one mutation so I've added a test to solve that too.

PHP 8 has also removed pcntl (used for timing) from core due to lack of maintainers. I've added that (and removed the aforementioned `chmod`) among some other cleanup which I'll post in a PR to php8-xdebug shortly.

If there's anything that needs changing in that PR first, please let me know and I'll get to it asap - I'll hold pushes to this PR until that's merged in case the pcntl error ends up causing CI to fail

**Comment by ChrisB9** [comment]
I just merged it, so let's see whether this now works. Thanks for digging into this!

**Comment by dshoreman** [comment]
No worries, thanks for the fast merge! I'll get the commits pushed right now... :pray:

**Comment by dshoreman** [comment]
Forgot about infection requiring 7.4, so I've added `--ignore-platform-reqs` to those versions in .travis.yml. I did try removing it from nightly too, but spatie/phpunit-watcher still trips that with a `^7.2` requirement.

Tests themselves are running again, but there's still some work to get coverage/infection working on Travis. I'll dig into that some more either today/tomorrow and update the PR if I figure anything out.

7.2 tests seem to only be failing currently due to latest infection using newer syntax (guessing trailing commas). We could just disable the infection task under 7.2? If that sounds good, I can add it in while I'm playing with the coverage bits.

Turning out to be quite the adventure, but... I think we're getting there!

**Comment by ChrisB9** [comment]
> did try removing it from nightly too, but spatie/phpunit-watcher still trips that with a ^7.2 requirement.

For phpunit-watcher, I quickly created a pull request to their repository https://github.com/spatie/phpunit-watcher/pull/124 so that we can omit `--ignore-platform-reqs`

> 7.2 tests seem to only be failing currently due to latest infection using newer syntax (guessing trailing commas). We could just disable the infection task under 7.2? If that sounds good, I can add it in while I'm playing with the coverage bits.

I wish I could drop 7.2 support altogether but sadly phpsu and our agency needs it.
Which line is the culprit for those failing infections?

> Turning out to be quite the adventure, but... I think we're getting there!
That's great, I am sure we will get to the bottom of this

**Comment by dshoreman** [comment]
The issue in 7.2 isn't failing infections, but Infection PHP *itself* using syntax that's incompatible with 7.2:
```
$ composer infection

> vendor/bin/infection --threads=4 --only-covered --min-msi=100 --min-covered-msi=100 --ansi

PHP Parse error:  syntax error, unexpected ')' in /home/travis/build/phpsu/ShellCommandBuilder/vendor/infection/infection/src/Container.php on line 223

Parse error: syntax error, unexpected ')' in /home/travis/build/phpsu/ShellCommandBuilder/vendor/infection/infection/src/Container.php on line 223

Script vendor/bin/infection --threads=4 --only-covered --min-msi=100 --min-covered-msi=100 --ansi handling the infection event returned with error code 255
```

My first thought was infection simply wouldn't run, but seeing your PR to phpunit-watcher hinted at an `^A | ^B` syntax for Composer's version specifiers. That's *especially* useful now because I only just realised phpunit was also refusing to run on 7.2:

```
$ composer test

> vendor/bin/phpunit -c tests/phpunit.xml --testdox --color=always

This version of PHPUnit is supported on PHP 7.3 and PHP 7.4.

You are using PHP 7.2.27 (/home/travis/.phpenv/versions/7.2.27/bin/php).

Script vendor/bin/phpunit -c tests/phpunit.xml --testdox --color=always handling the test event returned with error code 1

The command "composer test" exited with 1.
```

I'll try updating the composer.json to accept either version for phpunit and infection; hopefully that should sort it without having to exclude 7.2 from the infection/phpunit runs.

**Comment by dshoreman** [comment]
Some 30 builds and one sleep later, Travis is *finally* happy across all 7.x jobs. Huzzah!

Still got some work to do for the nightly job to pass, but once it does I'll rebase to clean up the huge string of commits.

**Comment by ChrisB9** [comment]
@dshoreman I see you did work quite a lot on it this weekend :D 
Can you give me a notice once you are ready?

**Comment by dshoreman** [comment]
> @dshoreman I see you did work quite a lot on it this weekend :D

It looks like a lot but by the time it's rebased I'm sure it'll be much less :laughing: 

> Can you give me a notice once you are ready?

Sure, will do. Had planned to finish it this morning but something came up so if I don't get time tonight, could be next day or two.

**Comment by dshoreman** [comment]
@ChrisB9 I've removed `allow-failures` from nightly builds (and it's still green!) but we're still waiting on a couple dependencies before we can remove `--ignore-platform-reqs` entirely from composer.

It all seems to be working fine though, so I'll create a separate issue with the details on that.  
Think this is finally good to review! :tada:

File: .travis.yml
Comment (ChrisB9) [inline .travis.yml]: i'd say we limit this the only requirement we really want to disable
```suggestion
      env: COMPOSER_OPTS="--ignore-platform-req=php"
```

**Review by ChrisB9 — CHANGES_REQUESTED** [review]
Looks good so far
Thanks for all your effort here!

**Comment by dshoreman** [comment]
@ChrisB9 Good idea to limit the ignores, committed that now so touch wood it'll all pass soon and it won't flag anything else!

**Review by ChrisB9 — APPROVED** [review]

**Pull request merged**
